lib/spack/spack/build_systems/compiler.py

```diff
@@ -0,0 +1,154 @@
+# Copyright 2013-2023 Lawrence Livermore National Security, LLC and other
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+import itertools
+import os
+import re
+import sys
+from typing import Sequence
+
+import llnl.util.tty as tty
+from llnl.util.lang import classproperty
+
+import spack.compiler
+import spack.package_base
+
+
+def true():
+    return True
+
+
+class CompilerPackage(spack.package_base.PackageBase):
+    """A Package mixin for all common logic for packages that implement compilers"""
+
+    # metadata identifying this as a compiler for lmod
+    family = "compiler"
+
+    # TODO: how do these play nicely with other tags
+    tags = ["compiler"]
+
+    # Optional suffix regexes for searching for this type of compiler.
+    # Suffixes are used by some frameworks, e.g. macports uses an '-mp-X.Y'
+    # version suffix for gcc.
+    suffixes: Sequence[str] = [r"-.*"]
+
+    # Optional prefix regexes for searching for this compiler
+    prefixes: Sequence[str] = []
+
+    #: Compiler argument(s) that produces version information
+    #: If multiple arguments, the earlier arguments must produce errors when invalid
+    #: Must be a hashable type (e.g. tuple, rather than list)
+    version_argument = "-dumpversion"
```

Comment:
Maybe let's just use:
```suggestion
    version_arguments: List[str] = ["-dumpversion"]
```
so that the type is clear.

Reply:
Closing this one, new comment in https://github.com/spack/spack/pull/43464/files#r1574929065